File: problems/leetcode/150-interview/208-Implement-Trie-(Prefix-Tree).cpp
struct Node{
map<char,Node*> children;
bool isWord;
Node(){
    isWord=false;
}
};
class Trie {
private:
Node* root;
public:
    Trie() {
        root=new Node();
    }
    
    void insert(string word) {
        Node* tmpPtr=root;
        int i=0;
        while(i<word.size()){
            if(tmpPtr->children.find(word[i])==tmpPtr->children.end()){
                Node* newChild= new Node();
                tmpPtr->children.insert({word[i],newChild});
                tmpPtr=newChild;
            }
            else{
                    tmpPtr=tmpPtr->children[word[i]];
            }
            i++;
        }
        tmpPtr->isWord=true;
    }
    
    bool search(string word) {
        Node* tmpPtr=root;
        int i=0;
        while(i<word.size()){
            if(tmpPtr->children.find(word[i])==tmpPtr->children.end()){
                return false;
            }
            else{
                    tmpPtr=tmpPtr->children[word[i]];
            }
            i++;
        }
        return tmpPtr->isWord==true;
    }
    
    bool startsWith(string prefix) {
        Node* tmpPtr=root;
        int i=0;
        while(i<prefix.size()){
            if(tmpPtr->children.find(prefix[i])==tmpPtr->children.end()){
                return false;
            }
            else{
                    tmpPtr=tmpPtr->children[prefix[i]];
            }
            i++;
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: problems/leetcode/LeetCode-75/208-Implement-Trie-(Prefix-Tree).py
class Trie:

    def __init__(self):
        self.children = {}

    def insert(self, word: str) -> None:
        current = self.children
        for each in word+'$':
            if each not in current:
                current[each] = {}
            current = current[each]

    def search(self, word: str) -> bool:
        return self.startsWith(word+'$')

    def startsWith(self, prefix: str) -> bool:
        current = self.children
        for each in prefix:
            if each not in current:
                return False
            current = current[each]
        return True

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)